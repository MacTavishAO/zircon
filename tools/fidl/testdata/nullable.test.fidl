library fidl.test.nullable;

using zx;

struct StructWithNullableString {
    string? val;
};

struct StructWithNullableVector {
    vector<int32>? val;
};

struct StructWithNullableHandle {
    zx.handle:VMO? val;
};

protocol SimpleProtocol {
    Add(int32 a, int32 b) -> (int32 sum);
};

struct StructWithNullableProtocol {
    SimpleProtocol? val;
};

struct StructWithNullableRequest {
    request<SimpleProtocol>? val;
};

struct Int32Wrapper {
    int32 val;
};

struct StructWithNullableStruct {
    Int32Wrapper? val;
};

union SimpleUnion {
    1: int32 a;
    2: float32 b;
};

struct StructWithNullableUnion {
    SimpleUnion? val;
};
