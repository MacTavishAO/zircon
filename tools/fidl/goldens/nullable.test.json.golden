{
  "version": "0.0.1",
  "name": "fidl.test.nullable",
  "library_dependencies": [
    {
      "name": "zx",
      "declarations": {
        "zx/CHANNEL_MAX_MSG_BYTES": "const",
        "zx/CHANNEL_MAX_MSG_HANDLES": "const",
        "zx/MAX_NAME_LEN": "const",
        "zx/MAX_CPUS": "const",
        "zx/obj_type": "enum",
        "zx/handle": "experimental_resource",
        "zx/status": "type_alias",
        "zx/time": "type_alias",
        "zx/duration": "type_alias",
        "zx/ticks": "type_alias",
        "zx/koid": "type_alias",
        "zx/vaddr": "type_alias",
        "zx/paddr": "type_alias",
        "zx/paddr32": "type_alias",
        "zx/gpaddr": "type_alias",
        "zx/off": "type_alias",
        "zx/procarg": "type_alias",
        "zx/signals": "type_alias"
      }
    }
  ],
  "bits_declarations": [],
  "const_declarations": [],
  "enum_declarations": [],
  "experimental_resource_declarations": [],
  "interface_declarations": [
    {
      "name": "fidl.test.nullable/SimpleProtocol",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 17,
        "column": 10,
        "length": 14
      },
      "methods": [
        {
          "ordinal": 1238508415331800925,
          "name": "Add",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 18,
            "column": 5,
            "length": 3
          },
          "has_request": true,
          "maybe_request": [
            {
              "type": {
                "kind": "primitive",
                "subtype": "int32"
              },
              "name": "a",
              "location": {
                "filename": "nullable.test.fidl",
                "line": 18,
                "column": 15,
                "length": 1
              },
              "field_shape_v1": {
                "offset": 16,
                "padding": 0
              }
            },
            {
              "type": {
                "kind": "primitive",
                "subtype": "int32"
              },
              "name": "b",
              "location": {
                "filename": "nullable.test.fidl",
                "line": 18,
                "column": 24,
                "length": 1
              },
              "field_shape_v1": {
                "offset": 20,
                "padding": 0
              }
            }
          ],
          "maybe_request_payload": "fidl.test.nullable/SomeLongAnonymousPrefix0",
          "maybe_request_type_shape_v1": {
            "inline_size": 24,
            "alignment": 8,
            "depth": 0,
            "max_handles": 0,
            "max_out_of_line": 0,
            "has_padding": false,
            "has_flexible_envelope": false,
            "is_resource": false
          },
          "has_response": true,
          "maybe_response": [
            {
              "type": {
                "kind": "primitive",
                "subtype": "int32"
              },
              "name": "sum",
              "location": {
                "filename": "nullable.test.fidl",
                "line": 18,
                "column": 37,
                "length": 3
              },
              "field_shape_v1": {
                "offset": 16,
                "padding": 4
              }
            }
          ],
          "maybe_response_payload": "fidl.test.nullable/SomeLongAnonymousPrefix1",
          "maybe_response_type_shape_v1": {
            "inline_size": 24,
            "alignment": 8,
            "depth": 0,
            "max_handles": 0,
            "max_out_of_line": 0,
            "has_padding": true,
            "has_flexible_envelope": false,
            "is_resource": false
          },
          "is_composed": false
        }
      ]
    }
  ],
  "service_declarations": [],
  "struct_declarations": [
    {
      "name": "fidl.test.nullable/SomeLongAnonymousPrefix0",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 18,
        "column": 8,
        "length": 18
      },
      "anonymous": true,
      "members": [
        {
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "name": "a",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 18,
            "column": 15,
            "length": 1
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        },
        {
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "name": "b",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 18,
            "column": 24,
            "length": 1
          },
          "field_shape_v1": {
            "offset": 4,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 8,
        "alignment": 8,
        "depth": 0,
        "max_handles": 0,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/SomeLongAnonymousPrefix1",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 18,
        "column": 30,
        "length": 11
      },
      "anonymous": true,
      "members": [
        {
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "name": "sum",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 18,
            "column": 37,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 4
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 8,
        "alignment": 8,
        "depth": 0,
        "max_handles": 0,
        "max_out_of_line": 0,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableString",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 5,
        "column": 8,
        "length": 24
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "string",
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 6,
            "column": 13,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 16,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableVector",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 9,
        "column": 8,
        "length": 24
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "vector",
            "element_type": {
              "kind": "primitive",
              "subtype": "int32"
            },
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 10,
            "column": 20,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 16,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 4294967295,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableHandle",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 13,
        "column": 8,
        "length": 24
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "handle",
            "subtype": "vmo",
            "rights": 2147483648,
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 14,
            "column": 20,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 4,
        "alignment": 4,
        "depth": 0,
        "max_handles": 1,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": true
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableProtocol",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 21,
        "column": 8,
        "length": 26
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.nullable/SimpleProtocol",
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 22,
            "column": 21,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 4,
        "alignment": 4,
        "depth": 0,
        "max_handles": 1,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": true
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableRequest",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 25,
        "column": 8,
        "length": 25
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "request",
            "subtype": "fidl.test.nullable/SimpleProtocol",
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 26,
            "column": 30,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 4,
        "alignment": 4,
        "depth": 0,
        "max_handles": 1,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": true
      }
    },
    {
      "name": "fidl.test.nullable/Int32Wrapper",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 29,
        "column": 8,
        "length": 12
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 30,
            "column": 11,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 4,
        "alignment": 4,
        "depth": 0,
        "max_handles": 0,
        "max_out_of_line": 0,
        "has_padding": false,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableStruct",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 33,
        "column": 8,
        "length": 24
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.nullable/Int32Wrapper",
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 34,
            "column": 19,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 8,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    },
    {
      "name": "fidl.test.nullable/StructWithNullableUnion",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 42,
        "column": 8,
        "length": 23
      },
      "anonymous": false,
      "members": [
        {
          "type": {
            "kind": "identifier",
            "identifier": "fidl.test.nullable/SimpleUnion",
            "nullable": true
          },
          "name": "val",
          "location": {
            "filename": "nullable.test.fidl",
            "line": 43,
            "column": 18,
            "length": 3
          },
          "field_shape_v1": {
            "offset": 0,
            "padding": 0
          }
        }
      ],
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    }
  ],
  "table_declarations": [],
  "union_declarations": [
    {
      "name": "fidl.test.nullable/SimpleUnion",
      "location": {
        "filename": "nullable.test.fidl",
        "line": 37,
        "column": 7,
        "length": 11
      },
      "members": [
        {
          "ordinal": 1,
          "reserved": false,
          "name": "a",
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "location": {
            "filename": "nullable.test.fidl",
            "line": 38,
            "column": 14,
            "length": 1
          }
        },
        {
          "ordinal": 2,
          "reserved": false,
          "name": "b",
          "type": {
            "kind": "primitive",
            "subtype": "float32"
          },
          "location": {
            "filename": "nullable.test.fidl",
            "line": 39,
            "column": 16,
            "length": 1
          }
        }
      ],
      "strict": true,
      "type_shape_v1": {
        "inline_size": 24,
        "alignment": 8,
        "depth": 1,
        "max_handles": 0,
        "max_out_of_line": 8,
        "has_padding": true,
        "has_flexible_envelope": false,
        "is_resource": false
      }
    }
  ],
  "type_alias_declarations": [],
  "declaration_order": [
    "fidl.test.nullable/StructWithNullableVector",
    "fidl.test.nullable/StructWithNullableUnion",
    "fidl.test.nullable/StructWithNullableStruct",
    "fidl.test.nullable/StructWithNullableString",
    "fidl.test.nullable/StructWithNullableRequest",
    "fidl.test.nullable/StructWithNullableProtocol",
    "fidl.test.nullable/StructWithNullableHandle",
    "fidl.test.nullable/SimpleProtocol",
    "fidl.test.nullable/SimpleUnion",
    "fidl.test.nullable/Int32Wrapper"
  ],
  "declarations": {
    "fidl.test.nullable/SimpleProtocol": "interface",
    "fidl.test.nullable/SomeLongAnonymousPrefix0": "struct",
    "fidl.test.nullable/SomeLongAnonymousPrefix1": "struct",
    "fidl.test.nullable/StructWithNullableString": "struct",
    "fidl.test.nullable/StructWithNullableVector": "struct",
    "fidl.test.nullable/StructWithNullableHandle": "struct",
    "fidl.test.nullable/StructWithNullableProtocol": "struct",
    "fidl.test.nullable/StructWithNullableRequest": "struct",
    "fidl.test.nullable/Int32Wrapper": "struct",
    "fidl.test.nullable/StructWithNullableStruct": "struct",
    "fidl.test.nullable/StructWithNullableUnion": "struct",
    "fidl.test.nullable/SimpleUnion": "union"
  }
}
